core/base/inc/TDirectory.h

```diff
@@ -139,6 +139,8 @@ can be replaced with the simpler and exception safe:
    TUUID            fUUID;              // Unique identifier
    mutable TString  fPathBuffer;        //! Buffer for GetPath() function
    TContext        *fContext{nullptr};  //! Pointer to a list of TContext object pointing to this TDirectory
+   /// Flag to signal whether the directory accepts registrations from histograms, graphs, etc.
+   Bool_t          fAcceptsRegisters{true};
```

Comment:
Isn't this redundant with `IsOpen()` ?

Reply:
IsOpen tells whether the file is open or not, `fAcceptsRegisters` tells whether a certain file accepts registrations from objects (of any type, but this should be checked by the different objects when trying to register themselves in the file)

Reply:
Is this feature (Open-but-rejecting-registration) related and/or necessary for this PR? (If not, the necessary discussion on this feature might detract from the original purpose of the PR).

Reply:
It is both related and necessary. But I agree the discussion should be moved elsewhere. I'll open a new PR for this :+1:

Reply:
> and necessary

Why :) ? The usual way to accomplish this (See TDirectory::TContext) is to temporarily set `gDirectory` (which is thread local) to `nullptr`.

Reply:
In the bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tfile.py of this PR there are two functions `_TFileEnter` and `_TFileExit` which will be set to the `__enter__` and `__exit__` dunder methods of the pythonized TFile. Both functions are needed in order to enable TFile usage as a context manager, i.e.
```python
with ROOT.TFile("myfile.root", "recreate") as myfile:
    h = ROOT.TH1F("myname","mytitle", 42, 13, 77)
    myfile.WriteObject(h, "myhisto")
```
for writing or
```python
with ROOT.TFile("myfile.root", "read") as myfile:
    h = myfile.myhisto
```
for reading to/from a TFile. Bear in mind that in Python, variables defined within a `with` context are kept alive and accessible after the context ends, e.g.
```python
with open("myfile.txt", "w") as myfile:
    myfile.write("Hello world\n")
    mylist = [1, 2, 3]
    myfile.write(f"This is a list {mylist}\n")

print("Info about the file")
print(f"{type(myfile)=}")
print(f"{myfile=}")
print(f"{bool(myfile)=}")
print(f"{myfile.closed=}")

print("\nInfo about the list object")
print(f"{type(mylist)=}")
print(f"{mylist=}")
print(f"{bool(mylist)=}")
```
results in 
```python
Info about the file
type(myfile)=<class '_io.TextIOWrapper'>
myfile=<_io.TextIOWrapper name='myfile.txt' mode='w' encoding='UTF-8'>
bool(myfile)=True
myfile.closed=True

Info about the list object
type(mylist)=<class 'list'>
mylist=[1, 2, 3]
bool(mylist)=True
```
But currently, the histograms written or read from the TFile inside the `with` context are nonified after the context, since they are automatically registered with the TFile and as soon as it is closed, all objects are gone too.

You approach of temporarily setting `gDirectory` to `nullptr` would then look like this:
```python
def _TFileEnter(obj):
    import ROOT
    ROOT._old_gDirectory = ROOT.TDirectory.CurrentDirectory().load() # copied from TContext
    ROOT.gDirectory = ROOT.nullptr
    return obj


def _TFileExit(obj, exc_type, exc_val, exc_tb):
    import ROOT
    obj.Close() # restore the directory after closing the file
    ROOT.gDirectory = ROOT._old_gDirectory
```

To me it already seems kind of hacky (notice the need to type `import ROOT` twice separately in the two functions because this Python file cannot `import ROOT` at module level (it's not initialised yet when the file is imported). Unfortunately, I tried and it doesn't seem to work. When writing the histogram, I can still access it after the `with`, but when reading it back it's nonified anyway. I guess when reading an object from a TFile, the fact that `gDirectory` is `nullptr` doesn't really matter?

Reply:
> I guess when reading an object from a TFile, the fact that gDirectory is nullptr doesn't really matter?

Right.

If I understand correctly, the main goal is to have the object survive the 'Exit' ... in this case why isn't the solution 
"just" to not close the `TFile`?

Reply:
On the one hand, the goal of the PR is to enable TFile as context manager behaving, like people would expect, equally to the direct Python counterpart, which I described above (i.e. the file is closed after the `with` ends).  
On the other hand, maybe we can still provide users with the `with` syntax and making it behave like what happens in Python *after* the context ends, i.e. objects are still alive. This option trades unbalance on the behaviour of the file with balance on the behaviour of the objects. I think we need opinions of @Axel-Naumann , @eguiraud , @etejedor about this alternative route

Reply:
Not closing the TFile at the end of the `with` block would be surprising for the user imo, since opening files with Python in a `with` block implies closing at the end. Also, if we don't close the TFile, why having a `with` block in the first place? It wouldn't do anything, would it?

Reply:
It could do (essentially):
```
enter: Open file and cd to file
exit: cd back to previous file/directory (including nullptr)
```
i.e. making sure it does not affect subsequent operations.

Note: that if you close the `TFile` the `TTree` (and `RNTuple`) objects are useless (basically this helps "only" histograms, `TGraph2D`, `TEntryList` and `TEventList`) [ well unless you also load all the `TTree` baskets in memory ]

Reply:
The `TTree` case is indeed special. All other cases still deserve an attempt at making this work. I'll retry again, with some suggestions from Axel, to restrict the changes to pythonizations without touching the C++ implementation of TFile.